integration/automation-extensions/src/main/java/org/wso2/esb/integration/common/extensions/carbonserver/CarbonServerManager.java

```diff
@@ -238,9 +238,21 @@ public synchronized void serverShutdown(int portOffset, boolean skipRunningCodeC
             log.info("Shutting down server ...");
 
             try {
+                /* Trying to stop the server for multiple times until it gets properly shutdown. Assuming server takes
+                around 5s to stop the server, waiting for 180s in total until the server gets shut down */
+                int retryCount = 36;
+                for (int i = 0; i < retryCount; i++) {
+                    if (isRemotePortInUse("localhost", managementPort)) {
+                        startProcess(carbonHome, getStartScriptCommand("stop"));
+                        waitTill(() -> isRemotePortInUse("localhost", managementPort), 5, TimeUnit.SECONDS);
```

Comment:
shouldn't we wait till !isRemotePortInUse?

Reply:
if the port is still in use is the server has not shut down yet, isn't it?

Reply:
so basically what waitTill() method does is it does a TimeUnit.SECONDS.sleep(1) until the first predicate which is isRemotePortInUse("localhost", managementPort) is true. Once it is false, waitTill() method will stop waiting and continue.